### PR TITLE
convox 3.0.14

### PR DESCRIPTION
**Pull request by ddollar**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/convox.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 class Convox < Formula
-  desc "Command-line interface for the Rack PaaS on AWS"
+  desc "Command-line interface for the Convox PaaS"
   homepage "https://convox.com/"
-  url "https://github.com/convox/rack/archive/20200129210159.tar.gz"
-  sha256 "945d6237ea0108ef0cd5ab9e6dae0f898710ca84a65d8480b8466e4c40ff99af"
+  url "https://github.com/convox/convox/archive/3.0.0.tar.gz"
```
Comment (SMillerDev) [inline Formula/convox.rb]: Since this is a very different versioning scheme you should add a line `version_scheme 1`

File: Formula/convox@2.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+class Convox < Formula
```
Comment (SMillerDev) [inline Formula/convox@2.rb]: This won't work. In addition to this, will v2 be supported by upstream for a while next to v3?

File: Formula/convox@2.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  desc "Command-line interface for the Rack PaaS on AWS"
+  homepage "https://convox.com/"
+  url "https://github.com/convox/rack/archive/20200129210159.tar.gz"
+  sha256 "945d6237ea0108ef0cd5ab9e6dae0f898710ca84a65d8480b8466e4c40ff99af"
+
+  bottle do
```
Comment (SMillerDev) [inline Formula/convox@2.rb]: Don't add bottles to new formulae.

**Review by SMillerDev — CHANGES_REQUESTED** [review]
Please submit the old version as a separate PR and make sure it complies with all rules in docs.brew.sh/Versions

**Comment by ddollar** [comment]
Thanks @SMillerDev ! I have updated the PR with your suggested changes.

**Review by zbeekman — CHANGES_REQUESTED** [review]
@ddollar: It looks like you have addressed all of @SMillerDev's concerns, and this is looking good so far.

I don't mean to be a stickler, but would it be possible to add a more meaningful test than [this](https://github.com/Homebrew/homebrew-core/pull/50398/files#diff-21e11853ad8eda1a46081b01de2b029bR39)?

If there's nothing obvious because you need to connect to authenticated web services or need a heavy docker install for testing, etc. you can always run a command that you expect to fail, match the output to check it fails in the way you expect and that the exit status is not 0 (if it shouldn't be zero).

Looking at the documentation though, it looks like you might be able to do something like check the output of `convox racks` or `convox instances`.

But without a better test the release may appear to work, when something more subtle is broken. If there's anything else that can be done (following my suggestions above) to verify expected behavior it would be much appreciated.

If this is not possible or unreasonable then please explain why.

Thanks so much!

**Comment by ddollar** [comment]
Thanks for the review @zbeekman 

Unfortunately all of the commands (including the ones you mentioned) require authentication with a remove API. I have modified the test to test the version output of `convox -v` is as expected which should validate that the CLI is working as much as can be without access to a remote API.

**Comment by zbeekman** [comment]
> Unfortunately all of the commands (including the ones you mentioned) require authentication with a remote API.

Could you test that the attempted authentication fails? You could even point it to a non-existent endpoint. That way you can see it is attempting to connect/authenticate and assert that it will fail. Would that be possible at all?

**Comment by zbeekman** [comment]
BTW, your current test is better than it was, but it would be nice if you could rebase and fixup---i.e., squash into a single commit---so if you could check that it attempts to authenticate and fails (with the expected output) that would be an improvement.

**Comment by ddollar** [comment]
Rebased and squashed. I gave some thought to going the auth/failed route but it seems like it would make the test flaky and dependent on local machine state (is the user already logged in?)

**Comment by zbeekman** [comment]
> I gave some thought to going the auth/failed route but it seems like it would make the test flaky and dependent on local machine state (is the user already logged in?)

Yes, as far as I know the user is already logged in. When you open up the CI VMs you can see the desktop etc. But you do need to take care that you have a mechanism to prevent timeouts and cleanup after the test, especially if you spawn an asynchronous process. There are some formulae that do this, but I'm trying to recall which ones and what to search on, but I'm drawing a blank. I know formulae that start services and servers, like `oauth2_proxy` will start a server in a forked process, sleep for a bit, then query them. And then others like `ddclient` show  examples of forking a process and cleaning up for client software connecting with remote servers.

Basically, run something like:
```
find "$(brew --repo homebrew/core)/Formula" -type f -name '*ctl*'
# or
find "$(brew --repo homebrew/core)/Formula" -type f -name '*client*'
```
To find examples of how other people do this. E.g., `ddclient`, `glooctl`, `fluxctl` etc.

Some folks are using simple/bad version checks, but whenever possible we try to get them to update. If you try something and CI barfs you can always revert it if you can't figure out how to make it work.

File: Formula/convox.rb [hunk 4]
```diff
@@ -19,21 +20,22 @@ class Convox < Formula
...
     end
 
-    system "go", "build", "-ldflags=-X main.version=#{version}",
-           "-o", bin/"convox", "-v", "github.com/convox/rack/cmd/convox"
     prefix.install_metafiles
   end
 
   test do
-    system bin/"convox"
+    assert_equal version, shell_output("#{bin}/convox -v").chomp
```
Comment (SMillerDev) [inline Formula/convox.rb]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

**Comment by ddollar** [comment]
Updated test to attempt invalid login

File: Formula/convox.rb [hunk 5]
```diff
@@ -19,21 +20,27 @@ class Convox < Formula
...
     end
 
-    system "go", "build", "-ldflags=-X main.version=#{version}",
-           "-o", bin/"convox", "-v", "github.com/convox/rack/cmd/convox"
     prefix.install_metafiles
   end
 
   test do
-    system bin/"convox"
+    require "open3"
```
Comment (SMillerDev) [inline Formula/convox.rb]: This is not needed, there's a homebrew dsl to read the shell output. `shell_output "<cmd>", <exitcode>`

Comment (ddollar) [reply Formula/convox.rb]: shell_output did not appear to capture stderr

Comment (ddollar) [reply Formula/convox.rb]: I was going off of the information found here: https://www.rubydoc.info/github/Homebrew/brew/Formula.test

Comment (SMillerDev) [reply Formula/convox.rb]: > shell_output did not appear to capture stderr

Not by default but you can just use `cmd 2>&1` to do that.

Comment (ddollar) [reply Formula/convox.rb]: The test seems more accurate as is

Comment (SMillerDev) [reply Formula/convox.rb]: The test also depends on gems that not everyone might have installed and is out of line with homebrew standards like this. Just use the homebrew DSL, it's there for a reason.

Comment (ddollar) [reply Formula/convox.rb]: Open3 is part of the Ruby standard library

https://www.rubydoc.info/stdlib/open3/Open3.popen3

Comment (SMillerDev) [reply Formula/convox.rb]: That doesn't change the homebrew standards. It's not really up for debate, it's not getting merged like this.

**Comment by zbeekman** [comment]
@ddollar The changes are looking good. I need to research what the official stance on `popen3` is vs capturing standard error and standard out with homebrew DSL. (Sean knows better than I do on this front so he's probably correct...) Also, if the build isn't queued by morning I'll check in on the CI system and if needed re-queue it. Feel free to ping me tomorrow if there's still no link to a Jenkins build.

**Review by zbeekman — CHANGES_REQUESTED** [review]
@ddollar please make the test so that this does not require open3. You can use shell_output and redirect standard out/stderr or `IO.popen`.

Thanks. Once this is done this is ready to merge.

**Comment by SMillerDev** [comment]
Didn't notice you pushed the changes, could you reword the commit to `<name> <version>`? Then it'll be good to go.

**Comment by SMillerDev** [comment]
And please tag me after, force-push is silent

**Comment by ddollar** [comment]
@SMillerDev updaetd

**Comment by SMillerDev** [comment]
Thanks @ddollar! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉
